pkg/controller/utils/gcp/utils.go

```diff
@@ -0,0 +1,174 @@
+)
+
+const (
+	infrastructureConfigName = "cluster"
+)
+
+var (
+	// Regexp to catch the first field of a permissions of the form
+	// <service>.<category>.<action> so that we can easily create
+	// the API name <service>.googleapis.com.
+	permToAPIRegexp = regexp.MustCompile(`^([A-Za-z0-9-]*)\..*`)
+)
+
+func init() {
+	if err := minterv1.AddToScheme(credentailRequestScheme); err != nil {
+		panic(err)
+	}
+}
+
+// CheckCloudCredCreation will see whether we have enough permissions to create new sub-creds, and
+// whether the necessary services are enabled.
+func CheckCloudCredCreation(gcpClient ccgcp.Client, logger log.FieldLogger) (bool, error) {
+	return checkServicesAndPermissions(gcpClient, CredMintingPermissions, logger)
+}
+
+// CheckCloudCredPassthrough will see if the provided creds are good enough to determine at
+// runtime whether the current credentials are good enough to be passed along as-is to satisfy
+// a CredentialsRequest (and validating that the associated APIs are enabled).
+func CheckCloudCredPassthrough(gcpClient ccgcp.Client, logger log.FieldLogger) (bool, error) {
+	return checkServicesAndPermissions(gcpClient, CredPassthroughPermissions, logger)
+}
+
+func checkServicesAndPermissions(gcpClient ccgcp.Client, permissionsList []string, logger log.FieldLogger) (bool, error) {
+	allowed, err := CheckPermissionsAgainstPermissionList(gcpClient, permissionsList, logger)
+	if err != nil {
+		return false, err
+	}
+
+	if !allowed {
+		return allowed, nil
+	}
+
+	servicesEnabled, err := CheckServicesEnabled(gcpClient, permissionsList, logger)
+	if err != nil {
+		return false, err
+	}
+
+	return servicesEnabled, nil
+}
+
+// CheckPermissionsAgainstPermissionList will take the passsed-in list of permissions to check whether the provided
+// gcpClient creds have sufficient permissions to perform the actions.
+// Will return true/false indicating whether the permissions are sufficient.
+func CheckPermissionsAgainstPermissionList(gcpClient ccgcp.Client, permList []string, logger log.FieldLogger) (bool, error) {
+	projectName := gcpClient.GetProjectName()
+
+	permRequest := &cloudresourcemanager.TestIamPermissionsRequest{
+		Permissions: permList,
+	}
+	permResponse, err := gcpClient.TestIamPermissions(projectName, permRequest)
+	if err != nil {
+		return false, fmt.Errorf("error testing permissions: %v", err)
+	}
+
+	// check that each perm in our list is actually available
+	permMap := map[string]bool{}
+	for _, perm := range permResponse.Permissions {
+		permMap[perm] = true
+	}
+
+	disallowedPerms := []string{}
+	for _, perm := range permList {
+		if _, ok := permMap[perm]; !ok {
+			disallowedPerms = append(disallowedPerms, perm)
+		}
+	}
+
+	if len(disallowedPerms) > 0 {
+		logger.Warn("Detected some unallowed permissions: %s", disallowedPerms)
+	}
+
+	return len(disallowedPerms) == 0, nil
+
+}
+
+// CheckServicesEnabled will take a list of GCP permissions, and see whether each permissions'
+// related API is enabled.
```

Comment:
What does it mean for APIs to be disabled in this context? Is this for a particular user and something in addition to permission checking?

Reply:
GCP has this idea of enabling/disabling individual service APIs at the project level. I can have complete admin permissions on a project, and still not have API access to different services until the API is enabled.

```
[jdiaz@minigoomba ~]$ gcloud spanner instances list
API [spanner.googleapis.com] not enabled on project

[jdiaz@minigoomba ~]$ gcloud services list --available --format='table(config.name,state)' --filter='config.name=spanner.googleapis.com'
NAME                    STATE
spanner.googleapis.com  DISABLED
```